docs/contribute/addnode.qmd

```diff
@@ -0,0 +1,170 @@
+
+- add `new_node_type::NewNodeType` to the Parameters object;
+- add `new_node_type = NewNodeType(db,config)` to the function `Parameters` in create.jl and add new_node_type add the proper location in the `Parameters` constructor call.
+
+## Reading from configuration
+
+To define a schema for the new node type, add the following to `validation.jl`:
+
+```julia
+@version NewNodeTypeV1 begin
+	node_id::Int
+	# Other fields
+end
+```
+
+Now we define the function that is called in the second bullet above, in `create.jl`:
+
+```julia
+function NewNodeType(db::DB, config::Config)::NewNodeType
+	static = load_structvector(db, config, NewNodeTypeV1)
+
+	# Unpack the fields of static as inputs for the NewNodeType constructor
+	return NewNodeType(static.node_id, static.some_property)
+end
+```
+
+## Node behaviour
+
+The behaviour of the new node in the Ribasim core is defined in a method of the `formulate!` function, which is called within the `water_balance!` (both in `solve.jl`) function being the right hand side of the system of differential equations solved by Ribasim. Here the details depend highly on the specifics of the node type. An example structure of a `formulate!` method is given below.
+
+```julia
+function formulate!(new_node_type::NewNodeType, p::Parameters)::Nothing
+	# Retrieve relevant parameters
+	(; connectivity) = p
+	(; flow) = connectivity
+	(; node_id, param_1, param_2) = new_node_type
+
+	# Loop over nodes of NewNodeType
+	for (i, id) in enumerate(node_id)
+		# compute e.g. flow based on param_1[i], param_2[i]
+	end 
+
+	return nothing
+end 
+```
+
+
+# Python I/O
+
+## Python class
+
+Create a new file `python/ribasim/ribasim/new_node_type.py` which is structured as follows:
+
+```python
+import pandas as pd
+import pandera as pa
+from pandera.typing import DataFrame, Series
+from pydantic import baseModel
+
+from ribasim.input_base import InputMixin
+
+__all__ = ("NewNodeType",)
+
+class StaticSchema(pa.SchemaModel):
+    node_id: Series[int] = pa.Field(coerce=True)
+    # Series per node property, pa.Field for mandatory data
+
+# Possible other schemas
+
+
+class NewNodeType(InputMixin, BaseModel):
+    """
+    Description of this node type.
+
+    Parameters
+    ----------
+    static: pandas.DataFrame
+        table with data for this node type.
+
+    possible other schemas
+    """
+
+    _input_type = "NewNodeType"
+    static: Optional[DataFrame[StaticSchema]] = None
+    # possible other schemas
+
+    class Config:
+        validate_assignment = True
+
+```
+
+in `python/ribasim/ribasim/model.py`, add
+
+- `from ribasim.new_node_type import NewNodeType`;
+- `(NewNodetype, "new_node_type")` in the `_NODES` list;
+- new_node_type as a parameter and in the docstring of the `Model` class.
+
+
+# QGIS plugin
+
+The script `qgis\core\nodes.py` has to be updated to specify how the new node type is displayed by the QGIS plugin. Specifically:
```

Comment:
```suggestion
The script `qgis/core/nodes.py` has to be updated to specify how the new node type is displayed by the QGIS plugin. Specifically:
```

We generally use forward slashes since they work better across OSes.